### PR TITLE
Optimise long multiply + add/sub/neg on arm64.

### PR DESCRIPTION
**Pull request by c272**
This PR optimises an extending multiply of two integers out to a long followed by an addition, subtraction or negation operation down to a single `[s/u]maddl`, `[s/u]subl` or `smnegl` instruction. It also adds a generic ternary operator which can be used in the future to support other three argument nodes.

The `spmidiff` results for this patch can be found at the below gist (performed on `win-arm64` since `linux-arm64` is currently broken as per #91257):
https://gist.github.com/c272/944721d2d083660e0ba7ec9fddb24de4

Minor regressions seen here are due to additional `mov` instructions changing the register allocation in cases where the add value is containable, preventing some previously performed `ldp` optimisations. Eliminating all cases of containable adds also removes many valid optimisations, so I have currently left it as is.

Contributes to https://github.com/dotnet/runtime/issues/68028

### CODE REVIEW

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -2521,6 +2521,84 @@ void CodeGen::genCodeForMulHi(GenTreeOp* treeNode)
     genProduceReg(treeNode);
 }
 
+//------------------------------------------------------------------------
+// genCodeForMulAddSubLong: Generates code for int*int + long -> long.
```
Comment (kunalspathak) [inline src/coreclr/jit/codegenarm64.cpp:2525]: also include an example for `SUB`.

**Review by kunalspathak — COMMENTED** [review]
It seems adding `GenTreeTernOp` can be avoided. In the past, we directly recognized the pattern in codegen e.g. `madd`, `msub`. Did you try that approach and faced any problem?

**Comment by a74nh** [comment]
> It seems adding `GenTreeTernOp` can be avoided. In the past, we directly recognized the pattern in codegen e.g. `madd`, `msub`. Did you try that approach and faced any problem?

I guess that would be possible, but it feels like it should be done during lowering (I have no overly strong opinion though).

My thought for `GenTreeTernOp` is that a follow on patch could move `SELECT`, `CMPXCHG` over to use it, which should simplify things a little.

**Comment by tannergooding** [comment]
Rather than `GenTreeTernOp`, we already have `GenTreeHWIntrinsic` which supports multi-op and which is really oriented around these types of specialized instructions for both SIMD and Scalar code.

We can just introduce an `ArmBase_LongMultiplyAdd` or similar for internal use only and have lowering generate that node. The existing metadata table will make this very simple to support and likely require less overall code changes.

**Comment by c272** [comment]
@tannergooding Thank you for the comment, I have reworked the patch to instead utilise `GT_HWINTRINSIC` instead of creating a new ternary node. The updated patch should also resolve the testing failures.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 2]
```diff
@@ -486,6 +486,14 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 break;
             }
 
+            case NI_ArmBase_Arm64_MultiplyLongAdd:
+                ins = node->TypeIs(TYP_ULONG) ? INS_umaddl : INS_smaddl;
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:490]: should this be `varTypeIsUnsigned(intrin.baseType)`?

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 3]
```diff
@@ -1112,6 +1120,13 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 GetEmitter()->emitIns_R_R_R(ins, emitSize, targetReg, op2Reg, op3Reg, opt);
                 break;
             }
+
+            case NI_ArmBase_Arm64_MultiplyLongAdd:
+            case NI_ArmBase_Arm64_MultiplyLongSub:
+                assert(opt == INS_OPTS_NONE);
+                GetEmitter()->emitIns_R_R_R_R(ins, EA_8BYTE, targetReg, op1Reg, op2Reg, op3Reg);
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:1127]:
```suggestion
                GetEmitter()->emitIns_R_R_R_R(ins, emitSize, targetReg, op1Reg, op2Reg, op3Reg);
```

File: src/coreclr/jit/lowerarmarch.cpp [hunk 4]
```diff
@@ -532,6 +532,16 @@ GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* binOp)
                 return next;
             }
         }
+
+        if (binOp->OperIs(GT_SUB))
+        {
+            // Attempt to optimise for smsubl.
```
Comment (kunalspathak) [inline src/coreclr/jit/lowerarmarch.cpp:538]:
```suggestion
            // Attempt to optimize for umsubl/smsubl.
```

File: src/coreclr/jit/lowerarmarch.cpp [hunk 5]
```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
         }
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// TryLowerAddSubToCombinedMulOp: Attempt to convert ADD and SUB nodes to a combined multiply
+// and add/sub operation, extending. Conversion can only happen if the operands to the
```
Comment (kunalspathak) [inline src/coreclr/jit/lowerarmarch.cpp:2717]:
```suggestion
// and add/sub operation. Conversion can only happen if the operands to the
```

File: src/coreclr/jit/lowerarmarch.cpp [hunk 6]
```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
...
+        addVal = op1;
+    }
+    else
+    {
+        // Exit if neither operation are GT_MUL_LONG.
+        return nullptr;
+    }
+
+    // Additional value must be of long size.
+    // The compiler will check the relevant signage for us, so we don't need to worry about that.
```
Comment (kunalspathak) [inline src/coreclr/jit/lowerarmarch.cpp:2776]: Not sure if I understood this comment. Can you please elaborate?

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Need to resolve merge conflict and also fix the test failure:

```
Assert failure(PID 275 [0x00000113], Thread: 275 [0x0113]): Assertion failed 'compiler->compIsaSupportedDebugOnly(HWIntrinsicInfo::lookupIsa(intrin.id))' in 'System.Collections.Generic.Dictionary`2[System.__Canon,System.__Canon]:TryInsert(System.__Canon,System.__Canon,ubyte):ubyte:this' during 'Generate code' (IL size 668; hash 0x26ab038f; FullOpts)

    File: /__w/1/s/src/coreclr/jit/hwintrinsiccodegenarm64.cpp Line: 210
    Image: /root/helix/work/correlation/corerun
````

File: src/coreclr/jit/lowerarmarch.cpp [hunk 7]
```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
...
+
+    // Mul values must both be integers.
+    if (!genActualTypeIsInt(mul->gtOp1) || !genActualTypeIsInt(mul->gtOp2))
+        return nullptr;
+
+    // Create the new node and replace the original.
+    {
+        NamedIntrinsic intrinsicId =
+            op->OperIs(GT_ADD) ? NI_ArmBase_Arm64_MultiplyLongAdd : NI_ArmBase_Arm64_MultiplyLongSub;
+        GenTreeHWIntrinsic* outOp = comp->gtNewScalarHWIntrinsicNode(mul->IsUnsigned() ? TYP_ULONG : TYP_LONG,
```
Comment (SingleAccretion) [inline src/coreclr/jit/lowerarmarch.cpp:2788]: > mul->IsUnsigned() ? TYP_ULONG : TYP_LONG

It is illegal to type nodes as `TYP_ULONG`.

Comment (SingleAccretion) [reply src/coreclr/jit/lowerarmarch.cpp:2788]: > add an assert to make it self-explanatory?

Definitely - there is a long-term desire to implement an "IR verifier" component (we already have a lot of the pieces, but they're a bit ad-hoc) that would check for things like that. It just hasn't been done yet.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 6 above)
Comment (c272) [reply src/coreclr/jit/lowerarmarch.cpp]: Honestly, I'm also not sure what I was trying to get at here. It was probably to denote that we didn't need to check for `TYP_ULONG`, but I have removed the comment line.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 7 above)
Comment (c272) [reply src/coreclr/jit/lowerarmarch.cpp]: Thank you, wasn't aware of this. I've fixed it as `TYP_LONG` and used the intrinsic's base type to determine unsigned/signed.

**Review by tannergooding — APPROVED** [review]
This looks correct to me. It should get sign off from someone from @dotnet/jit-contrib before being merged

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**